hubblestack/extmods/modules/nebula_osquery.py

```diff
@@ -1390,32 +1391,100 @@ def _parse_log(path_to_logfile,
     else:
         log.error("Log file doesn't exists: {0}".format(path_to_logfile))
 
     return event_data
 
-
-def _set_cache_offset(path_to_logfile, offset):
+def _set_file_offset(path_to_logfile, offset):
     '''
-    Cache file offset in memory
+    Cache file offset in specified file
+    A file will be created in cache directory and following attributes will be stored in it
+    offset, initial_crc (CRC for first 256 bytes of log file), last_crc (CRC for last 256 bytes of log file)
     '''
-    cachefilekey = os.path.basename(path_to_logfile)
-    __RESULT_LOG_OFFSET__[cachefilekey] = offset
-
+    try:
+        log_filename = os.path.basename(path_to_logfile)
+        offsetfile = os.path.join(__opts__.get('cachedir'), 'osqueryd', 'offset', log_filename)
+        log_file_initial_crc = 0
+        log_file_last_crc = 0
+        if(offset > 0):
+            with open(path_to_logfile, 'r') as log_file:
+                log_file.seek(0)
+                log_file_initial_crc = zlib.crc32(log_file.read(CRC_BYTES))
+
+            if(offset > CRC_BYTES):
+                with open(path_to_logfile, 'r') as log_file:
+                    log_file.seek(offset - CRC_BYTES)
+                    log_file_last_crc = zlib.crc32(log_file.read(CRC_BYTES))
+
+        offset_dict = {"offset" : offset, "initial_crc": log_file_initial_crc, "last_crc": log_file_last_crc}
+        log.info("Storing following information for file {0}. Offset: {1}, Initial_CRC: {2}, Last_CRC: {3}".format(path_to_logfile, offset, log_file_initial_crc, log_file_last_crc))
+        if not os.path.exists(os.path.dirname(offsetfile)):
+            os.makedirs(os.path.dirname(offsetfile))
+
+        with open(offsetfile, 'w') as json_file:
+            json.dump(offset_dict, json_file)
+    except Exception as e:
+        log.error("Exception in creating offset file. Exception: {0}".format(e))
+        tb = traceback.format_exc()
+        log.error("Exception stacktrace: {0}".format(tb))
 
 def _get_file_offset(path_to_logfile):
     '''
-    Fetch file offset for specified file
+    Fetch file offset for specified file.
+    It works by loading JSON offset file stored in cache directory.
+    From the JSON file, it fetches offset, initial_crc and last_crc of the log file.
+    Return 0 if there is any exception
     '''
-    cachefilekey = os.path.basename(path_to_logfile)
-    offset = __RESULT_LOG_OFFSET__.get(cachefilekey, 0)
+    offset = 0
+    try:
+        log_filename = os.path.basename(path_to_logfile)
+        offsetfile = os.path.join(__opts__.get('cachedir'), 'osqueryd', 'offset', log_filename)
+        if not os.path.isfile(offsetfile):
+            log.info("Offset file: {0} does not exist. Returning offset as 0.".format(offsetfile))
+        else:
+            with open(offsetfile, 'r') as file:
+                offset_data = json.load(file)
+            offset = offset_data['offset']
```

Comment:
fetching values in this form from a dictionary is error prone. This will throw an exception if 'offset' key is not found, and could break Hubble's workflow. Instead fetch with **offset_data.get('offset')**. Similar for other places.

Reply:
fixed this issue. Good to know.